### PR TITLE
Fix/661/additional attributes

### PR DESCRIPTION
**Pull request by nc-fkl**
- Added multiple attributes
- Aligned mapping
- Aligned frontend layout a littlebit so that its more categorized
- Setting newly added properties of account via IAccountManager instead of IUser

### CODE REVIEW

**Review by julien-nc — CHANGES_REQUESTED** [review]
Nice!
Unit and integration tests don't pass.
I'll make real life tests when they do.

File: lib/Service/ProvisioningService.php [hunk 1]
```diff
@@ -67,19 +74,72 @@ public function provisionUser(string $tokenUserId, int $providerId, object $idTo
...
+		$headline = $idTokenPayload->{$headlineAttribute} ?? null;
+
+		$biographyAttribute = $this->providerService->getSetting($providerId, ProviderService::SETTING_MAPPING_BIOGRAPHY, 'biography');
+		$biography = $idTokenPayload->{$biographyAttribute} ?? null;
+
 		$event = new AttributeMappedEvent(ProviderService::SETTING_MAPPING_UID, $idTokenPayload, $tokenUserId);
 		$this->eventDispatcher->dispatchTyped($event);
 
-		$backendUser = $this->userMapper->getOrCreate($providerId, $event->getValue());
+		$backendUser = $this->userMapper->getOrCreate($providerId, (string)$event->getValue());
```
[the remark this answers is not in the document]
Comment (nc-fkl) [reply lib/Service/ProvisioningService.php:135]: Actualy my ide marked it as error that's not casted, that was the only reason for that :) took your solution now

File: lib/Service/ProvisioningService.php [hunk 2]
```diff
@@ -127,6 +188,115 @@ public function provisionUser(string $tokenUserId, int $providerId, object $idTo
...
+				$country
+			];
+		}
+
+		// filter out empty values
+		$addressParts = array_filter($addressParts, function($value) {
+			return !empty($value);
+		});
+		// concatenate them back together into a string and remove unsused ', '
+		$address = preg_replace('/^, |, $/', '', str_replace('  ', ' ', implode(', ', $addressParts)));
```
[the remark this answers is not in the document]
Comment (nc-fkl) [reply lib/Service/ProvisioningService.php:223]: Ah good point, makes sense, so lets keep the "ugly" string when a value was not provided, will also be a good hint for the user/admin itself that a value isn't set correctly.

File: lib/Command/UpsertProvider.php [hunk 3]
```diff
@@ -176,10 +193,60 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		if ($mapping = $input->getOption('mapping-uid')) {
 			$this->providerService->setSetting($provider->getId(), ProviderService::SETTING_MAPPING_UID, $mapping);
 		}
+		if ($mapping = $input->getOption('mapping-fediverse')) {
+			$this->providerService->setSetting($provider->getId(), ProviderService::SETTING_MAPPING_FEDIVERSE, $mapping);
+		}
+		if ($mapping = $input->getOption('mapping-groups')) {
+			$this->providerService->setSetting($provider->getId(), ProviderService::SETTING_MAPPING_GROUPS, $mapping);
+		}
```
Comment (julien-nc) [inline lib/Command/UpsertProvider.php:201]: You forgot to add this option in the configure method. This will make the integration tests pass.

**Review by julien-nc — CHANGES_REQUESTED** [review]
Check the unit tests output. They need to be adjusted with the new setting values.

**Review by julien-nc — CHANGES_REQUESTED** [review]
Nice, i tried with a few attributes, works like a charm.

It would be nice to hide all those setting fields and make them appear/disappear if user clicks on a "Extra  attributes mapping" button.

Also, now there are conflicts with main. Nothing weird, the new options in the UpSert command are added on the same line. Let me know if you need help to solve the conflicts.

**Comment by nc-fkl** [comment]
> Nice, i tried with a few attributes, works like a charm.
> 
> It would be nice to hide all those setting fields and make them appear/disappear if user clicks on a "Extra attributes mapping" button.
> 
> Also, now there are conflicts with main. Nothing weird, the new options in the UpSert command are added on the same line. Let me know if you need help to solve the conflicts.

Redbased and added expandable/collapsable extra attributes. Would be nice if you can review again 🏓 , thank you @julien-nc :)

**Review by julien-nc — APPROVED** [review]
Woop woop

**Pull request merged**